### PR TITLE
docs-issue-gen: Fix testing failures

### LINKED ISSUE #110682: docs-issue-generation: Epic extraction test flakes
https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_BazelEssentialCi/11780610?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildTestsSection=true&expandBuildProblemsSection=true&expandBuildDeploymentsSection=true

Seems like the test is failing. Need to spend some time debugging. This could be related to DEVINF-884, but it's unclear at the moment.

Jira issue: CRDB-31555

### PR DESCRIPTION
**Pull request by nickvigilante**
Fixes #110682 #110704

Release note: none

### CODE REVIEW

**Review by healthy-pod — COMMENTED** [review]
can you explain how this deflakes the test?

**Comment by nickvigilante** [comment]
@healthy-pod Sure:

Within [`extractEpicIDs`](https://github.com/cockroachdb/cockroach/blob/6e4cb1cc9145de2bd8138a33176f2f04c480311b/pkg/cmd/docs-issue-generation/extract.go#L91), we extract a list of potential epic IDs from a message (in our case, the commit body concatenated with the PR body) via the `extractStringsFromMessage`. I found that some engineers were putting tickets that were **not** epics while still using the `Epic:` prefix, so I needed to add logic to validate that the ticket in the epic itself is actually an epic, or if it's not, return the epic to which that ticket belongs, or return nothing for the epic reference.

The [second test for `TextExtractEpicIDs`](https://github.com/cockroachdb/cockroach/blob/6e4cb1cc9145de2bd8138a33176f2f04c480311b/pkg/cmd/docs-issue-generation/docs_issue_generation_test.go#L3037) has a series of epic references in the body, including a mapping from `https://cockroachlabs.atlassian.net/browse/CRDB-28708` to `CRDB-18955`.

In the `extractEpicIDs` function, we build up the `invalidEpicRefs` global map to prevent duplicate calls to the network when possible. Observe how it's built over time:

```
result: map[CRDB-235:1 CRDB-9234:1 DEVINF-392:1 DOC-6883:1 https://cockroachlabs.atlassian.net/browse/CRDB-28708:1]
---
issueKey: CRDB-31495
invalidEpicRefs: map[]
epicKey: CRDB-31495
isEpic: true
---
issueKey: CRDB-31495
invalidEpicRefs: map[CRDB-31495:CRDB-31495]
epicKey: CRDB-31495
isEpic: true
---
issueKey: CRDB-491
invalidEpicRefs: map[CRDB-31495:CRDB-31495]
epicKey: CRDB-491
isEpic: true
---
issueKey: https://cockroachlabs.atlassian.net/browse/CRDB-28708
invalidEpicRefs: map[CRDB-31495:CRDB-31495 CRDB-491:CRDB-491]
epicKey: CRDB-18955
isEpic: false
--- # start of the problem; the `for` loop is updated include `CRDB-18955` because of line 117: https://github.com/cockroachdb/cockroach/blob/6e4cb1cc9145de2bd8138a33176f2f04c480311b/pkg/cmd/docs-issue-generation/extract.go#L117
issueKey: CRDB-18955
invalidEpicRefs: map[CRDB-31495:CRDB-31495 CRDB-491:CRDB-491 https://cockroachlabs.atlassian.net/browse/CRDB-28708:CRDB-18955]
epicKey:
isEpic: false
--- # end of the problem
issueKey: CRDB-9234
invalidEpicRefs: map[CRDB-31495:CRDB-31495 CRDB-491:CRDB-491 https://cockroachlabs.atlassian.net/browse/CRDB-28708:CRDB-18955]
epicKey:
isEpic: false
---
issueKey: CRDB-235
invalidEpicRefs: map[CRDB-31495:CRDB-31495 CRDB-491:CRDB-491 https://cockroachlabs.atlassian.net/browse/CRDB-28708:CRDB-18955]
epicKey: CRDB-235
isEpic: true
---
issueKey: DOC-6883
invalidEpicRefs: map[CRDB-235:CRDB-235 CRDB-31495:CRDB-31495 CRDB-491:CRDB-491 https://cockroachlabs.atlassian.net/browse/CRDB-28708:CRDB-18955]
epicKey: DOC-6883
isEpic: true
---
issueKey: DEVINF-392
invalidEpicRefs: map[CRDB-235:CRDB-235 CRDB-31495:CRDB-31495 CRDB-491:CRDB-491 DOC-6883:DOC-6883 https://cockroachlabs.atlassian.net/browse/CRDB-28708:CRDB-18955]
epicKey:
isEpic: false
```

As we can see from the above output, we check `https://cockroachlabs.atlassian.net/browse/CRDB-28708`, and the epic key we're given back is `CRDB-18955`. However, because `CRDB-18955` is not in `invalidEpicRefs` and the monkey-patched resources don't include a value, the current logic of the function removes `CRDB-18955` from the list, because it doesn't think it's an epic.

There are two solutions to this problem:

1. When we add `https://cockroachlabs.atlassian.net/browse/CRDB-28708: CRDB-18955` to the `invalidEpicRefs` map, also add `CRDB-18955: CRDB-18955` to the map alongside it.
2. Add a value to the result map of the monkey-patched test to ensure when we check `CRDB-18955`, it always knows it's an epic for the purposes of testing.

After writing up this response, I've come to realize that even though I implemented approach 2 in this PR, I will also want to incorporate approach 1 as well, since 2 only benefits testing, but 1 benefits both testing and the actual execution of the script.

**Comment by nickvigilante** [comment]
Update: further findings show that iterating over a map with `range` while simultaneously adding elements to that same map in that same `for` loop will occasionally iterate over the added elements. It's this iteration over a newly-added element that's part of the problem.

File: pkg/cmd/docs-issue-generation/extract.go [hunk 1]
```diff
@@ -90,33 +91,40 @@ func extractInformIssueIDs(message string) map[string]int {
...
 		}
 		if isEpic {
 			continue
-		} else if issueKey != epicKey {
+		} else if key != epicKey {
 			if epicKey != "" {
 				result[epicKey] = count
 			}
-			delete(result, issueKey)
+			delete(result, key)
```
Comment (rickystewart) [inline pkg/cmd/docs-issue-generation/extract.go:127]: I don't know to what extent it was affecting the problem before, but it is definitely a big improvement to not mutate the `result` `map` as we are iterating through it. That can very well introduce the kind of non-determinism we have been struggling with.

**Review by rickystewart — APPROVED** [review]

File: pkg/cmd/docs-issue-generation/extract.go (hunk 1 above)
Comment (nickvigilante) [reply pkg/cmd/docs-issue-generation/extract.go:127]: I'm better off because of the struggle to figure this out. Thanks for your assistance!

**Pull request merged**